### PR TITLE
feat: onboard dynamic yield

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

Onboard dynamic yield

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  - path: ../../bindings/jsontemplate
+
+steps:
+  - name: messageType
+    template: |
+      .message.type.toLowerCase()
+  - name: validateInput
+    template: |
+      let messageType = $.outputs.messageType;
+      $.assert(.message.type, "message Type is not present. Aborting message.");
```
Comment (koladilip) [inline src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:16]: should we do validation before messageType step

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      const userId = $.outputs.commonFields.user.id;
+      const sessionId = $.outputs.commonFields.session.custom;
+      const ip = $.outputs.commonFields.context.device.ip;
+      $.assert(userId, "userId is required");
+      $.assert(sessionId, "sessionId is required");
+      $.assert(ip, "ip is required");
+
+  - name: prepareIdentifyPayload
+    condition: $.outputs.messageType === {{$.EventType.IDENTIFY}}
+    template: |
```
Comment (koladilip) [inline src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:49]: this is not required

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+          hashedEmail = hashedEmail? sha256(hashedEmail): hashedEmail;
+          let properties = {
+            "hashedEmail": hashedEmail,
+            "dyType": 'identify-v1'
+          };
+          properties = $.removeUndefinedAndNullValues(properties);
+      - name: updatedPropertiesFields
+        condition: $.outputs.prepareIdentifyPayload.propertiesFields.hashedEmail === null
+        template: |
+          let updatedProperties = $.outputs.prepareIdentifyPayload.propertiesFields;
```
Comment (koladilip) [inline src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:66]: why do we need this if we are assigning  in line 68

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          let cuid = $.outputs.commonFields.user.id;
+          updatedProperties = {
+            "cuidType": "userId",
+            "cuid": cuid
+          };
+      - name: events
+        template: |
+          let properties = $.outputs.updatedPropertiesFields? $.outputs.updatedPropertiesFields : $.outputs.propertiesFields;
+          let events = {
```
Comment (koladilip) [inline src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:75]: use [{}], it is more efficient in this case

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml (hunk 1 above)
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:15]: Done

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml (hunk 2 above)
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml (hunk 3 above)
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  - name: prepareIdentifyPayload
+    condition: $.outputs.messageType === {{$.EventType.IDENTIFY}}
+    steps:
+      - name: propertiesFields
+        template: |
+          const email = .message.().(
+            {{{{$.getGenericPaths("email")}}}};
+            );
+          const hashEmail = .destination.Config.hashEmail;
+          hashedEmail = hashEmail ? SHA256(email) : email;
```
Comment (ujjwal-ab) [inline src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:60]: @koladilip Am I using it correctly? Can you please check once?

Comment (koladilip) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:60]: https://github.com/rudderlabs/rudder-workflow-engine/pull/67 check this you will get complete idea.

summary: you need to use like `$. SHA256(email)`

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml (hunk 4 above)
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml (hunk 5 above)
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+          properties = $.removeUndefinedAndNullValues(properties);
+      - name: updatedPropertiesFields
+        condition: $.outputs.prepareIdentifyPayload.propertiesFields.hashedEmail === undefined
+        template: |
+          let cuid = $.outputs.commonFields.user.id;
+          const updatedProperties = {
+            "dyType": 'identify-v1',
+            "cuidType": "userId",
+            "cuid": cuid
+          };
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:74]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      - name: events
+        template: |
+          let properties = $.outputs.prepareIdentifyPayload.updatedPropertiesFields? $.outputs.prepareIdentifyPayload.updatedPropertiesFields : $.outputs.prepareIdentifyPayload.propertiesFields;
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:77]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+          ...$.outputs.prepareTrackPayload.populateEventProperties,
+          "properties": {
+            ...$.outputs.prepareTrackPayload.populateEventProperties.properties,
+            "quantity" : .message.properties.quantity,
+            "productId": .message.properties.sku || .message.properties.product_id
+            }
+          }
+
+      - name: populateCartObj
+        condition: .message.properties.products && Array.isArray(.message.properties.products)
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            "cuidType": "userId",
+            "cuid": cuid
+          };
+      - name: events
+        template: |
+          let properties = $.outputs.prepareIdentifyPayload.updatedPropertiesFields? $.outputs.prepareIdentifyPayload.updatedPropertiesFields : $.outputs.prepareIdentifyPayload.propertiesFields;
+          let events = [{
+            "name": "Identify User",
+            "properties": properties 
+          }];
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    steps:
+      - name: getTrimmedEvent
+        template: |
+          const trimmedEvent = .message.event.trim().toLowerCase().split(" ").join("_");
+
+      - name: populateEventProperties
+        condition: $.outputs.prepareTrackPayload.getTrimmedEvent in ["product_added", "product_removed", "product_added_to_wishlist", "order_completed"]
+        template: |
+          const trimmedEvent = $.outputs.prepareTrackPayload.getTrimmedEvent;
+          const dyType = $.ecomEventsMapping[trimmedEvent].toLowerCase().split(" ").join("-")+"-v1";
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:94]: We are doing this for the value that we are getting from `$.ecomEventsMapping[trimmedEvent]`

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+              ...eventProperties,
+              cart: .message.properties@prop.products.({
+                "productId": .sku ?? prop.sku,
+                "quantity": Number(.quantity ?? prop.quantity),
+                "itemPrice": Number(parseFloat(.price ?? prop.price).toFixed(2))
+              })
+            }
+          }
+
+      - name: populateCustomEventProperties
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            "name": "Identify User",
+            "properties": properties 
+          }];
+
+  - name: prepareTrackPayload
+    condition: $.outputs.messageType === {{$.EventType.TRACK}}
+    steps:
+      - name: getTrimmedEvent
+        template: |
+          const trimmedEvent = .message.event.trim().toLowerCase().split(" ").join("_");
```
[the remark this answers is not in the document]
Comment (koladilip) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:88]: what about this? is it expected?

[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:88]: No, but using regex I tried like this. But it didn't work.
```
const mappedEvent = $.ecomEventsMapping[trimmedEvent].toLowerCase();
const dyType = mappedEvent.replaceAll(new RegExp('\\s+', 'g'),'-');a
```
Is there anything wrong in this?

[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Corrcted

**Review by koladilip — DISMISSED** [review]

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        "type": "REST",
+        "method": "POST",
+        "endpoint": $.ENDPOINT,
+        "headers": {
+          "DY-API-Key": .destination.Config.apiKey,
+          "Content-Type": "application/json"
+        },
+        "params": {},
+        "files": {}
+      }
```
Comment (ItsSudip) [inline src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml:188]: Check if we can use the shared defaultRequestConfig() here.

Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+  - name: messageType
+    template: |
+      .message.type.toLowerCase()
+
+  - name: validateInputForTrack
+    description: Additional validation for Track events
+    condition: $.outputs.messageType === {{$.EventType.TRACK}}
+    template: |
+      $.assert(.message.event, "event_name could not be mapped. Aborting.")
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+  - name: removeUndefinedAndNullValues
+    path: ../../../../v0/util
+  - path: ../../bindings/jsontemplate
+
+steps:
+  - name: validateInput
+    template: |
+      let messageType = .message.type;
+      $.assert(.message.type, "message Type is not present. Aborting message.");
+      $.assertConfig(.destination.Config.apiKey, "Invalid Api Key");
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

File: src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml [hunk 16]
```diff
@@ -0,0 +1,188 @@
+bindings:
+  - name: EventType
+    path: ../../../../constants
+  - path: ../../bindings/jsontemplate
+    exportAll: true
+  - path: ./config
+  - name: removeUndefinedAndNullValues
+    path: ../../../../v0/util
+  - path: ../../bindings/jsontemplate
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/cdk/v2/destinations/dynamic_yield/procWorkflow.yaml]: Updated

**Review by koladilip — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**